tensorflow/python/keras/engine/compile_utils_test.py

```diff
@@ -531,6 +531,13 @@ def test_accuracy(self):
     self.assertEqual(metric_container.metrics[0]._fn,
                      metrics_mod.binary_accuracy)
 
+    loss = losses_mod.BinaryCrossentropy()
+    metric_container = compile_utils.MetricsContainer('accuracy', loss=loss)
```

Comment:
@huonw Let me know if your case is covered by this test.

Reply:
Looks like it, but there is a reduced test case in #41361 that I cut down from our real code and so is the "true" measure of success. Does that test case match the behaviour of 2.1 with this PR?

Reply:
I've tried that code locally and It Is working with this patch. Can you verify it on your side or check if you want to extend this test coverage.

Reply:
Great, thanks! I'd be happy to say that this PR resolves the issue from our side then.

Unfortunately, I am not in a position to build TensorFlow from source, so unless there's an easy way for me to install a package from CI, I can't confirm on our real code myself. However, this issue was flagged in public code: in particular, https://github.com/stellargraph/stellargraph/issues/1766 found Jupyter notebook where the model reported accuracy metrics of 0 with TF 2.2 (but not TF 2.1). This can be run locally using something like:

1. Download the notebook from https://stellargraph.readthedocs.io/en/v1.2.1/demos/link-prediction/gcn-link-prediction.ipynb
2. Install necessary libraries `pip install stellargraph[demos]==1.2.1`
3. Run the notebook and see whether the `fit` call reports an accuracy of constant zero (bad) or decreasing non-zero (good)

This is a bit fiddly, so I'm not expecting you to do so. 😄

Reply:
>Unfortunately, I am not in a position to build TensorFlow from source, so unless there's an easy way for me to install a package from CI, 

I know that generally It Is resource intensive task and It could be a barrier (E.g. See https://github.com/tensorflow/build/issues/5).

If you see the diff in this PR  is very small so with a little "hack" you can just edit/copy these changes over you installed Tensorflow package. 
You can found where the installed files are with `pip show <your_installed_tensorflow_package_name>`.
Let me know.

Reply:
Ah, good trick. I applied it on Colab: https://colab.research.google.com/gist/huonw/4ac8796f598bc742b476fd8e36a1f866/gcn-link-prediction.ipynb . It didn't work.

Discussion (numbers `[n]` refer to the cell execution count on [the rendered form](https://nbviewer.jupyter.org/gist/huonw/4ac8796f598bc742b476fd8e36a1f866)):

1. I applied the patch with `curl ... | patch` [2]
2. I confirmed the patch was applied with `grep` [3]
3. I ran the example from #41361 with many ways to specify the loss [4]. For example, `"bce"` and referring to the `keras.losses.binary_crossentropy` function directly (the example was extended to include printing the `loss` value used). Output here for convenience:
   ```
   'bce' binary_accuracy
   'binary_crossentropy' categorical_accuracy
   <function binary_crossentropy at 0x7ff1b8db0950> categorical_accuracy
   <tensorflow.python.keras.losses.BinaryCrossentropy object at 0x7ff1a8dc1208> binary_accuracy
   <function binary_crossentropy at 0x7ff1b8db0950> categorical_accuracy
   ```
   Note that only `loss="bce"` and `loss=tf.keras.losses.BinaryCrossentropy()` work, `loss="binary_crossentropy"` and the two forms of `loss=tf.keras.losses.binary_crossentropy` do not.
4. I ran rest of the notebook from https://github.com/stellargraph/stellargraph/issues/1766 without changes to build and compile a Keras model [5]-[19] (using `loss=tf.keras.losses.binary_crossentropy`, as in the original)
5. I inspected the Keras model to confirm that it's using the `MetricsContainer` class with the `loss` parameter to its constructor and the `_loss` attribute [20], [21]
6. I evaluated and trained the model (also without changes from the original notebook), observing that the accuracy was the incorrect 0 value [22]-[25]